service/history/workflow/mutable_state_impl.go

```diff
@@ -4955,33 +4962,78 @@ func (ms *MutableStateImpl) updateActivityInfoForRetries(
 	nextAttempt int32,
 	activityFailure *failurepb.Failure,
 ) {
-	ms.updateActivity(ai, func(info *persistencespb.ActivityInfo, impl *MutableStateImpl) *persistencespb.ActivityInfo {
-		ai = UpdateActivityInfoForRetries(
-			ai,
-			ms.GetCurrentVersion(),
-			nextAttempt,
-			ms.truncateRetryableActivityFailure(activityFailure),
-			timestamppb.New(nextScheduledTime),
-		)
-		return ai
+	ms.UpdateActivityWithCallback(ai, func(info *persistencespb.ActivityInfo, mutableState MutableState) {
+		mutableStateImpl, ok := mutableState.(*MutableStateImpl)
+		if ok {
+			ai = UpdateActivityInfoForRetries(
+				ai,
+				mutableStateImpl.GetCurrentVersion(),
+				nextAttempt,
+				mutableStateImpl.truncateRetryableActivityFailure(activityFailure),
+				timestamppb.New(nextScheduledTime),
+			)
+		}
 	})
 }
 
-func (ms *MutableStateImpl) updateActivity(ai *persistencespb.ActivityInfo, updateCallback func(*persistencespb.ActivityInfo, *MutableStateImpl) *persistencespb.ActivityInfo) {
-	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity
-	// info and if prev found it points to the same activity info as ai, so updating ai will cause
-	// size of prev change.
+func (ms *MutableStateImpl) UpdateActivityWithCallback(
+	ai *persistencespb.ActivityInfo,
+	updateCallback UpdateActivityCallback,
+) {
+	// incoming activity info can be a pointer to the activity info in pendingActivityInfoIDs
+	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity info
+	// if prev found it can point to the same activity info as incoming activity info,
 	var originalSize int
 	if prev, ok := ms.pendingActivityInfoIDs[ai.ScheduledEventId]; ok {
 		originalSize = prev.Size()
 	}
+
 	updateCallback(ai, ms)
 
 	ms.approximateSize += ai.Size() - originalSize
 	ms.updateActivityInfos[ai.ScheduledEventId] = ai
 	ms.syncActivityTasks[ai.ScheduledEventId] = struct{}{}
 }
 
+func (ms *MutableStateImpl) UpdatePausedEntitiesSearchAttribute() error {
+	// TODO: search attributes are coupled with mutable state.
+	// This should be refactored so we don't add new function to MutableState interface every time we want to work with search attributes.
+
+	pausedInfoMap := make(map[string]struct{})
+
+	for _, ai := range ms.GetPendingActivityInfos() {
+		if !ai.Paused {
+			continue
+		}
+		pausedInfoMap[ai.ActivityType.Name] = struct{}{}
+	}
+	pausedInfo := make([]string, 0, len(pausedInfoMap))
+	for activityType := range pausedInfoMap {
+		activityType = url.QueryEscape(activityType)
+		pausedInfo = append(pausedInfo, fmt.Sprintf("activity::%s", activityType))
+	}
+	if len(pausedInfo) > 0 {
+		pausedInfo = append(pausedInfo, "policy::ManualActivityPause")
```

Comment:
This isn't a policy, this is a manual pause request, e.g: `reason:manual:%s` and let the user specify the reason so this can be used to find all affected workflows.

Reply:
I would put this in the loop that goes over all of the paused activities so you can make a more local decision on the value here while maintaining the set and dedupe semantics.

Reply:
lets wait

Reply:
remove until we better understand what we want